updated model record field guide (#672)
osohq/oso: docs/getting-started/application/patterns.rst

An actor can only approve an expense if they are the manager of the submitter
and the expense amount is below a certain limit.

Primary and Secondary Authorization Data
----------------------------------------

This policy controls access to an ``Expense``, the **primary authorization
data**.  It relies on other data to make the decision: the submitter of the
expense (``expense.submitted_by``), and the manager of the submitter
(``submitted_by.manager``).  This *other* data is called **secondary
authorization data**.  An important class of secondary authorization data is
**actor data**.  This data includes properties of the actor, like their role, or
team membership that is often used in controlling access regardless of whether
it is over rows, columns or fields.

Where the Policy is Evaluated

> -----------------------------

Reviewer: I actually think this content is useful. I don't think we need to make a fuss about defining terms like "policy evaluation point" unless we decide to use them consistently throughout the docs. But I do think it's worth touching on where the policy is evaluated. I think the best thing would be to have some kind of diagram that explains this better. E.g., a diagram showing the flow of the request and where the policy check is happening in this case.

> middleware.

Reviewer: This sentence is a little vague. Is this referring to some kind of request middleware?

on it's type. This is called **model level** can be used for coarse level authorization.
This is useful if you want to restrict access to a whole page if the user cannot see any
of the resources of the kind the page is about. Since the rules depend only on the type
of the model and not on the data it makes it easier to add these checks upstream in
middleware.

.. code-block:: polar
    :caption: :fa:`oso`

    allow(actor, "view", "expense") if actor.role = "accountant";
    allow(actor, "modify", "team") if actor.role = "hr_admin";
    allow(actor, "modify", "project") if actor.role = "hr_admin";
    allow(actor, "modify", "organization") if actor.role = "hr_admin";
    allow(actor, "view", "Expense") if actor.role = "accountant";
    allow(actor, "modify", "Team") if actor.role = "hr_admin";
    allow(actor, "modify", "Project") if actor.role = "hr_admin";
    allow(actor, "modify", "Organization") if actor.role = "hr_admin";

This brief policy shows an example of model level access control:

- An accountant can view expenses.
- HR admins can modify teams, projects, and organizations.

Notice that this policy does not rely on any **primary authorization data**.

> Therefore it can be evaluated either before or after the primary data fetch.

Reviewer: Think it might be worth keeping this in to reinforce the point, but replace "primary authorization data" with "resource attributes" or something like that.

> You can write a rule saying if a user can create an Expense or not.

Reviewer: It would be good to have more explanation here of why model-level access control is good for this.

> access control. Record level is about authorizing an action on a specific instance

Reviewer: ```suggestion
access control. Record level is about authorizing access to a specific instance
```

> of a resource. Is the user allowed to edit **this** Expense. Record level

Reviewer: ```suggestion
of a resource. It asks the question: is the actor allowed to edit **this** Expense? Record level
```

> save it to the database if it is.

Reviewer: This feels a little random without more set-up.

            # Not authorized.
            return NotAuthorizedResponse()

> Here, we were able to evaluate the policy **before data fetch** (modification in

Reviewer 2: Why did you want to remove this?

>     We use the ``oso.query`` method in this example to query a rule other than

Reviewer: We should link to the method for `oso.query`

    ``allow``.

.. todo::

    relevant link & this is incorrect with our API now!

In this example, we evaluated both record & column level access control after
data fetch.  However, it may be more efficient to use column level access
control to only load the columns the user can access:

.. code-block:: python
    :caption: :fab:`python`

    from oso import Variable

    def get_expense(user, expense_id):
        # Query oso for all fields allowed for this user.
        allowed_fields = oso.query_rule("allow_field",
                                   user, "view", expense, Variable("field"))
        # Convert the returned query response into a list of fields
        allowed_fields = [r["field"] for r in allowed_fields]
        allowed_fields_sql = db.sql_escape(allowed_fields.join(", "))

        expense = db.fetch(
            f"SELECT {allowed_fields_sql} FROM expenses WHERE id = %",

>             expense_id)

Reviewer: This is actually very cool but definitely out of place without more explanation of stuff like `Variable`.

        # Record level authorization.
        if oso.is_allowed(user, "view", expense):
            # Return only authorized_data to the user.
            ...
        else:
            # Not authorized
            return NotAuthorizedResponse()

Now, we are using oso to tell us what fields to query for.  In this example, the
policy is evaluated both **before and after data fetch** for greater efficiency.

.. admonition:: Variables provide flexibility

    Notice that we didn't have to change our policy file at all to make this
    change from the previous example. We passed in a ``Variable`` which
    instructs oso to find all values of ``field`` that match the rules we
    defined in our policy.  This flexibility derives directly from writing a
    :doc:`declarative policy in Polar </more/language/polar-foundations>`!

.. _authorizing-list-endpoints:

> Authorizing List Endpoints

Reviewer 2: I think we should keep most of this with maybe some wording updates and more context about how this compares to data filtering

Reviewer 2: We should probably split it into a separate page about data filtering
